usage_example/devbuf_map.c

```diff
@@ -0,0 +1,238 @@
+* @param virt_addr: Virtual address to convert
+* @param map_start_virt_addr: Starting virtual address of the mapped buffer
+* @param dev_map_offset: Offset of the mapped buffer in the device
+*
+* OUT: phys_addr
+* @param phys_addr: Pointer to store the physical address
+* 
+* @return: 0 on success, -1 on failure
+*/
+static int devbuf_virt_to_phys(int       buf_num, 
+                               void     *virt_addr, 
+                               void     *map_start_virt_addr,
+                               uint64_t  dev_map_offset,
+                               uint64_t *phys_addr)
+{
+    char dev_path[BUFSIZ];
+    int ret;
+    int fd;
+    uint64_t offset;
+    devbuf_details devbuf;
+
+    printf("devbuf_virt_to_phys: buf_num: %d, virt_addr: %p, map_start_virt_addr: %p, dev_map_offset: %lx\n", 
+            buf_num, virt_addr, map_start_virt_addr, dev_map_offset);
+    if (!phys_addr){
+        printf("Invalid phys_addr pointer\n");
+        return -1;
+    }
+
+    if (virt_addr < map_start_virt_addr){
+        printf("Invalid virt_addr, less than starting dev map address\n");
+        return -1;
+    }
+    offset = (uint64_t)virt_addr - (uint64_t)map_start_virt_addr + dev_map_offset;
+
+    ret = get_device_buffer_details(buf_num, &devbuf);
+    if (ret < 0) 
+        return -1;
+
+    if (offset > devbuf.size){
+        printf("Invalid offset, greater than buffer size\n");
+        return -1;
+    }
+
+    *phys_addr = devbuf.start_phys_addr + offset;
+    return 0;
+}
+
+
+/*
+* Map a device buffer to user space.
+*
+* IN: buf_num, size   
+* @param buf_num: Number of the device buffer
+* @param size: Size of the buffer to map
+*
+* OUT: map_addr, devbuf
+* @param map_addr: Pointer to store the mapped address
+* @param devbuf: Pointer to the devbuf_details structure to store the buffer details
+*   
+*/
+int map_device_buffer(unsigned int buf_num, size_t size, void **map_addr, devbuf_details *devbuf)
+{
+    char dev_path[BUFSIZ];
+	void *addr;
+    int fd;
+
+    snprintf(dev_path, sizeof(dev_path), "/dev/udmabuf%d", buf_num);
+    fd = checked_open(dev_path, O_RDWR);
+
+	if (fd != -1) {
+        printf("Getting device buffer details for %s\n", dev_path);
+        get_device_buffer_details_filedes(fd, devbuf);
+
+        printf("mapping device file %s\n", dev_path);
+        addr = mmap(NULL, size, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
+		*map_addr = addr;
+		close(fd);
+    }
+
+    return 0;
+}
+
+int main(int argc, char **argv)
+{
+    char buffer[BUFSIZ];
+    int maps_fd;
+	void *map_addr;
+	size_t buf_size;
+    devbuf_details devbuf;
+    uint64_t phys_addr;
+
+
+    if (argc < 2) {
+        printf("Usage: %s size (in MB)\n", argv[0]);
+        return EXIT_FAILURE;
+    }
+    buf_size = strtoull(argv[1], NULL, 0);
+	buf_size = buf_size << 20;
+
+	map_device_buffer(0, buf_size, &map_addr, &devbuf);
+	printf("map_addr: %p\n", map_addr);
+
+    devbuf_virt_to_phys(0, (map_addr + 4096), map_addr, 0, &phys_addr);
```

Comment:
What is the 4096 4K offset for ?

Reply:
Just an example of a virtual address, nothing special about 4096.